File: src/protocol.ts
// SPDX-License-Identifier: MIT

/** Metadata sent alongside the Markdown body in a save request. */
export interface SaveRequestMetadata {
  url: string;
  canonical_url: string;
  title: string;
  author?: string;
  site?: string;
  lang?: string;
  excerpt?: string;
  word_count?: number;
  saved_at: string;
  /** Default tags to apply on save. Ignored by host versions that don't support it. */
  tags?: string[];
}

/** The wire protocol version shared with the native host. */
export const PROTOCOL_VERSION = 1;

/**
 * One image the extension captured from the page (reusing the browser's cache),
 * base64-encoded so it can ride inside the JSON save message. The host decodes
 * and writes these; it never downloads anything itself.
 */
export interface ImageData {
  /** The URL exactly as it appears in the Markdown, so the host can rewrite it. */
  url: string;
  /** The response's `Content-Type`, used by the host to pick a file extension. */
  content_type: string;
  /** Standard base64 of the raw image bytes. */
  data_base64: string;
}

/** Message sent from the extension to the native host. */
export interface SaveRequest {
  protocol_version: typeof PROTOCOL_VERSION;
  action: "save";
  metadata: SaveRequestMetadata;
  markdown: string;
  images: ImageData[];
}

export interface SaveResponseSuccess {
  protocol_version: typeof PROTOCOL_VERSION;
  ok: true;
  id: string;
  path: string;
}

export interface SaveResponseError {
  protocol_version: typeof PROTOCOL_VERSION;
  ok: false;
  error: string;
  message: string;
}

export type SaveResponse = SaveResponseSuccess | SaveResponseError;

/** Message sent from the extension to check whether a URL is already saved. */
export interface CheckRequest {
  protocol_version: typeof PROTOCOL_VERSION;
  action: "check";
  url: string;
}

export interface CheckResponse {
  protocol_version: typeof PROTOCOL_VERSION;
  ok: true;
  saved: boolean;
  id?: string;
}

File: native-host/src/protocol.rs
// SPDX-License-Identifier: MIT

use serde::{Deserialize, Serialize};

/// Incoming save request from the browser extension.
#[derive(Debug, Deserialize)]
pub struct SaveRequest {
    pub protocol_version: u32,
    pub action: String,
    pub metadata: RequestMetadata,
    pub markdown: String,
    pub image_urls: Vec<String>,
}

#[derive(Debug, Deserialize)]
pub struct RequestMetadata {
    pub url: String,
    pub canonical_url: String,
    pub title: String,
    pub author: Option<String>,
    pub site: Option<String>,
    pub lang: Option<String>,
    pub excerpt: Option<String>,
    pub word_count: Option<u32>,
    pub saved_at: String,
}

/// Incoming check request — asks whether a URL is already in the library.
#[derive(Debug, Deserialize)]
pub struct CheckRequest {
    pub protocol_version: u32,
    pub url: String,
}

/// Outgoing response to the browser extension (shared by save and check actions).
#[derive(Debug, Serialize)]
pub struct SaveResponse {
    pub protocol_version: u32,
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    /// Present only in responses to a `check` action.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub saved: Option<bool>,
}

impl SaveResponse {
    pub fn success(id: String, path: String) -> Self {
        Self {
            protocol_version: 1,
            ok: true,
            id: Some(id),
            path: Some(path),
            error: None,
            message: None,
            saved: None,
        }
    }

    pub fn error(code: &str, msg: &str) -> Self {
        Self {
            protocol_version: 1,
            ok: false,
            id: None,
            path: None,
            error: Some(code.to_string()),
            message: Some(msg.to_string()),
            saved: None,
        }
    }

    pub fn check(is_saved: bool, id: Option<String>) -> Self {
        Self {
            protocol_version: 1,
            ok: true,
            saved: Some(is_saved),
            id,
            path: None,
            error: None,
            message: None,
        }
    }
}
